File: app/src/main/java/me/li2/android/architecture/ui/articledetail/view/ArticleDetailFragment.java
package me.li2.android.architecture.ui.articledetail.view;

import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.transition.TransitionInflater;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import javax.inject.Inject;

import butterknife.BindView;
import me.li2.android.architecture.R;
import me.li2.android.architecture.data.model.Article;
import me.li2.android.architecture.ui.articledetail.viewmodel.ArticleDetailViewModel;
import me.li2.android.architecture.ui.basic.BaseFragment;
import me.li2.android.architecture.utils.BaseImageLoader;
import me.li2.android.architecture.utils.ViewUtils;

public class ArticleDetailFragment extends BaseFragment {

    public static final String EXTRA_ARTICLE_ID = "article_id";

    @Inject
    ArticleDetailViewModel mViewModel;

    @Inject
    BaseImageLoader mImageLoader;

    @BindView(R.id.articleImageView)
    ImageView mImageView;

    @BindView(R.id.articleTitleView)
    TextView mTitleView;
    
    @BindView(R.id.articleDescriptionView)
    TextView mDescriptionView;

    public static ArticleDetailFragment newInstance(int articleId) {
        Bundle args = new Bundle();
        args.putInt(EXTRA_ARTICLE_ID, articleId);
        ArticleDetailFragment fragment = new ArticleDetailFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // shared element transition between RecyclerView and Fragment
        // tell the Fragment to wait to load until we tell it
        postponeEnterTransition();
        // tell it what type of Transition we want
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            setSharedElementEnterTransition(TransitionInflater.from(getContext()).inflateTransition(android.R.transition.move));
        }
    }

    @Override
    public int getLayout() {
        return R.layout.article_detail_fragment;
    }

    @Override
    public void setupView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
    }

    @Override
    public void onResume() {
        super.onResume();

        mViewModel.getArticle(getArticleId()).observe(this, article -> {
            if (article != null) {
                bindArticle(article);
            }
        });
    }

    private void bindArticle(Article article) {
        mTitleView.setText(article.getTitle());
        mDescriptionView.setText(article.getDescription());
        // shared element transition between RecyclerView and Fragment. notebyweiyi
        ViewUtils.setArticleTransitionName(mImageView.getContext(), mImageView, article.getTitle());
        mImageLoader.loadImage(mImageView, article.getImageHref(), null, succeed -> startPostponedEnterTransition());
    }

    private int getArticleId() {
        if (getArguments() != null) {
            return getArguments().getInt(EXTRA_ARTICLE_ID);
        }
        return -1;
    }
}

File: app/src/main/java/me/li2/android/architecture/ui/articledetail/view/ArticleDetailFragment.kt
package me.li2.android.architecture.ui.articledetail.view

import android.arch.lifecycle.Observer
import android.content.Context
import android.os.Build
import android.os.Bundle
import android.support.v4.view.ViewCompat
import android.transition.TransitionInflater
import android.view.View
import io.reactivex.functions.Consumer
import kotlinx.android.synthetic.main.article_detail_fragment.*
import me.li2.android.architecture.R
import me.li2.android.architecture.data.model.Article
import me.li2.android.architecture.ui.articledetail.viewmodel.ArticleDetailViewModel
import me.li2.android.architecture.ui.basic.BaseFragment
import me.li2.android.architecture.utils.BaseImageLoader
import javax.inject.Inject

private const val EXTRA_ARTICLE_ID = "article_id"

fun newArticleDetailFragmentInstance(articleId: Int): ArticleDetailFragment {
    val args = Bundle()
    args.putInt(EXTRA_ARTICLE_ID, articleId)
    val fragment = ArticleDetailFragment()
    fragment.arguments = args
    return fragment
}

fun setArticleTransitionName(context: Context, sharedElement: View, id: String?) {
    val transitionName = context.getString(R.string.transition_article_cover_image) + "." + id
    ViewCompat.setTransitionName(sharedElement, transitionName)
}


class ArticleDetailFragment : BaseFragment() {

    override val layout: Int = R.layout.article_detail_fragment

    @Inject
    lateinit var mViewModel: ArticleDetailViewModel

    @Inject
    lateinit var mImageLoader: BaseImageLoader

    private val articleId: Int
        get() {
            return if (arguments != null) arguments!!.getInt(EXTRA_ARTICLE_ID) else -1
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // shared element transition between RecyclerView and Fragment
        // tell the Fragment to wait to load until we tell it
        postponeEnterTransition()

        // tell it what type of Transition we want
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            sharedElementEnterTransition = TransitionInflater.from(context).inflateTransition(android.R.transition.move)
        }
    }

    override fun onResume() {
        super.onResume()

        mViewModel.getArticle(articleId).observe(this, Observer { article ->
            article?.let {
                bindArticle(it)
            }
        })
    }

    private fun bindArticle(article: Article) {
        articleTitleView.text = article.title
        articleDescriptionView.text = article.description
        // shared element transition between RecyclerView and Fragment. notebyweiyi
        setArticleTransitionName(context!!, articleImageView, article.title)
        mImageLoader.loadImage(articleImageView, article.imageHref, null, Consumer { startPostponedEnterTransition() })
    }
}
